cloudpickle/cloudpickle.py

```diff
@@ -249,6 +252,138 @@ def _walk_global_ops(code):
                 yield op, instr.arg
 
 
+def _memoryview_from_bytes(data_holder, format, readonly, shape):
+    """Build a memoryview from raw bytes read from the pickle stream
+
+    This function tries its best not to allocate any extra memory when not
+    necessary.
+
+    The actual data is provided in a data_holder which is expected to be a
+    single item list holding the actual bytes object. This makes it easier
+    to use sys.getrefcount to detect whether or not the bytes are referenced
+    elsewhere or not.
+
+    If readonly is False and the backing bytes in data_holder are referenced
+    from other objects, then the bytes contents are copied into a newly
+    allocated mutable buffer. Otherwise the memory allocated for the bytes
+    object is directly reused by the new memoryview.
+
+    Python 2 memoryviews doe not support being cast so that the
+    format and shape parameters are ignored.
+
+    Reference counting cannot be used in PyPy to detect if it is safe to mutate
+    the bytes object. If readonly is False, a new mutable buffer is always
+    allocated.
+    """
+    # If the following call to sys.getrefcount returns 2 it means that there is
+    # one reference from data_holder and one from the temporary arguments
+    # datastructure of the sys.getrefcount call. It is therefore safe to reuse
+    # the memory buffer of the bytes object as writeable buffer to back the
+    # memoryview: this buffer is no longer unreachable from anywhere else.
+    if hasattr(sys, 'getrefcount'):
+        safe_to_mutate = sys.getrefcount(data_holder[0]) <= 2
```

Comment:
When making calls like this, it may be nice to call out what the references are:
1. The reference owned by `data_holder`
2. The reference pushed onto the data stack from the expression `data_holder[0]`

Reply:
This is what I explained in the above comment. Also, I reorganized the code to make it more readable and easier to test.